SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoopInvariantCodeMotion.swift

```diff
@@ -0,0 +1,1020 @@
+      return false
+    }
+
+    if let callee = referencedFunction,
+       callee.hasSemanticsAttribute("array.props.isNativeTypeChecked") {
+      return false
+    }
+
+    return !calleeAnalysis.getSideEffects(ofApply: self).memory.write
+  }
+  
+  func isSafeReadOnlyApply(
+    for sideEffects: StackWithCount<Instruction>,
+    _ aliasAnalysis: AliasAnalysis,
+    _ calleeAnalysis: CalleeAnalysis
+  ) -> Bool {
+    if calleeAnalysis.getSideEffects(ofApply: self).memory == .noEffects {
+      return true
+    }
+
+    for sideEffect in sideEffects {
+      switch sideEffect {
+      case let storeInst as StoreInst:
+        if storeInst.storeOwnership == .assign ||
+           mayRead(fromAddress: storeInst.destination, aliasAnalysis) {
+          return false
+        }
+      case let copyAddrInst as CopyAddrInst:
+        if !copyAddrInst.isInitializationOfDestination ||
+           mayRead(fromAddress: copyAddrInst.destination, aliasAnalysis) {
+          return false
+        }
+      case is ApplyInst, is BeginApplyInst, is TryApplyInst:
+        if calleeAnalysis.getSideEffects(ofApply: self).memory.write {
+          return false
+        }
+      case is CondFailInst, is StrongRetainInst, is UnmanagedRetainValueInst,
+        is RetainValueInst, is StrongRetainUnownedInst, is FixLifetimeInst,
+        is KeyPathInst, is DeallocStackInst, is DeallocStackRefInst,
+        is DeallocRefInst:
+        break
+      default:
+        if sideEffect.mayWriteToMemory {
+          return false
+        }
+      }
+    }
+
+    return true
+  }
+}
+
+private extension BeginAccessInst {
+  func canBeHoisted(
+    outOf loop: Loop,
+    analyzedInstructions: AnalyzedInstructions,
+    _ context: FunctionPassContext
+  ) -> Bool {
+    guard endAccessInstructions.allSatisfy({ loop.contains(block: $0.parentBlock)}) else {
+      return false
+    }
+    
+    let areBeginAccessesSafe = analyzedInstructions.beginAccesses
+      .allSatisfy { otherBeginAccessInst in
+        guard self != otherBeginAccessInst else { return true }
+
+        return self.accessPath.isDistinct(from: otherBeginAccessInst.accessPath)
+      }
+
+    guard areBeginAccessesSafe else { return false }
+    
+    var scope = InstructionRange(begin: self, ends: endAccessInstructions, context)
+    defer { scope.deinitialize() }
+
+    for fullApplyInst in analyzedInstructions.fullApplies {
+      guard mayWriteToMemory
+          ? fullApplyInst.mayReadOrWrite(
+            address: address,
+            context.aliasAnalysis
+          ) : fullApplyInst.mayWrite(
+            toAddress: address,
+            context.aliasAnalysis
+          ) else {
+        continue
+      }
+
+      if !scope.contains(fullApplyInst) {
+        return false
+      }
+    }
+
+    switch accessPath.base {
+    case .class, .global:
+      for sideEffect in analyzedInstructions.loopSideEffects where sideEffect.mayRelease {
```

Comment:
I can't remember what's the reasoning behind this check. Can you add a comment explaining this?

Reply:
Right, thanks! Can you add this as a comment?